### DISCUSSION TITLE
Allow marking a class as virtual.

### DISCUSSION THREAD

**Discussion by moonheart08** [dotnet/csharplang#8068] [Language Ideas]
In short, allow requiring the programmer to be explicit about the usage for their class (inheritable and constructable, sealed and constructable, or inheritable and abstract), through a disabled-by-default warning.

This would enable the workflow I've seen in some repositories like https://github.com/space-wizards/space-station-14 where the usage of a class must be specified explicitly to be done without a custom analyzer and attribute (`[Virtual]` in the case of that repo)

(I'm entirely up for implementing this myself.)

**Comment by CyrusNajmabadi** [comment]
The attribute + analyzer route seems ideal here.  Esp. as it would not require a new lang version, and already works (virtually on every Roslyn compiler version out there).

**Comment by colejohnson66** [comment]
What is a "virtual" class? One that must be inherited to be used? Isn't that just an abstract class?

**Reply by moonheart08** [reply]
Same as a virtual field, it may or may not be inherited. It is, essentially, just an elaboration on a standard class definition to make sure the programmer is genuinely opting into that contractual obligation (inheritability with constructor) instead of just forgetting to mark it sealed or abstract.

**Reply by CyrusNajmabadi** [reply]
Could you clarify why the attribute+analyzer route is not desired here?  It seems to fit the exact use case of what we defined the analyzer system for.  Namely that you can institute specific rules to limit what would otherwise be allowed by the language.
